modules/stochastic_tools/src/state_sim/model/diagrams/StateSimDiagramEval.C

```diff
@@ -0,0 +1,88 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+
+#include "StateSimDiagramEval.h"
+#include "StateSimDiagram.h"
+#include <string>
+//#include "NextObjID.h"
+
+StateSimDiagramEval::StateSimDiagramEval(std::string name, DIAGRAM_TYPE_ENUM diag_type)
+  : StateSimDiagram(name, diag_type),
+    _diag_type(diag_type),
+    _ok_states(),
+    //_single_state_group(),
+    _fail_states()
+{
+  //_id = NextObjID::instance()->nextID(OID_STATE);
+  static int counter = 0;
+  _id = ++counter;
+}
+
+bool
+StateSimDiagramEval::evaluate(StateSimBitset cur_states, bool on_success)
+{
+  if (on_success)
+  {
+    for (auto it = _ok_states.begin(); it != _ok_states.end(); ++it)
+    {
+      if (cur_states[it->first])
+        return true;
+    }
+  }
+  else
+  {
+    for (auto it = _fail_states.begin(); it != _fail_states.end(); ++it)
+    {
+      if (cur_states[it->first])
+        return true;
+    }
+  }
+
+  //not found so return false
+  return false;
+}
+
+bool
+StateSimDiagramEval::isFailedState(int state_id)
+{
+  std::unordered_map<int, StateSimState *>::const_iterator got = _fail_states.find(state_id);
+  return got != _fail_states.end();
+}
+
+void
+StateSimDiagramEval::addState(StateSimState & add_state, const bool & is_success_state)
+{
+  _states[add_state.id()] = &add_state;
```

Comment:
Also, I prefer explicit inserts instead of bracket operators but these are OK if you like them for clarity or syntax.